src/core/ext/filters/client_channel/retry_filter.cc

```diff
@@ -932,7 +932,10 @@ void RetryFilter::CallData::CallAttempt::AddBatchesForPendingBatches(
       ++num_callbacks;
     }
     if (batch->recv_message) {
-      if (completed_recv_message_count_ < started_recv_message_count_) {
+      // Skip if the op is already in flight, or if it has already completed
+      // but the completion has not yet been sent to the surface.
+      if (completed_recv_message_count_ < started_recv_message_count_ ||
```

Comment:
would we start the op later in this case?

Reply:
We don't need to start the op later.  The bug here was that we were replaying the op on this call attempt when it had actually already been completed but the completion was not yet returned to the surface.  This prevents us from replaying it incorrectly.

Reply:
ah ok, got it